server.go

```diff
@@ -296,19 +305,53 @@ type Server struct {
 	DecorateReader DecorateReader
 	// DecorateWriter is optional, allows customization of the process that writes raw DNS messages.
 	DecorateWriter DecorateWriter
+	// Graceful shutdown handling
+	running int32
+	// Messages queue
+	queue chan *message
+	// Number of workers, if set to zero - use spawn goroutines instead
+	Workers int
+	// Size of DNS requests queue
+	QueueSize int
+}
+
+func (srv *Server) start() {
+	atomic.StoreInt32(&(srv.running), int32(1))
+}
+
+func (srv *Server) stop() {
+	atomic.StoreInt32(&(srv.running), int32(0))
+}
 
-	// Shutdown handling
-	lock    sync.RWMutex
-	started bool
+func (srv *Server) isRunning() bool {
+	return atomic.LoadInt32(&(srv.running)) == 1
+}
+
+func (srv *Server) startWorkers() {
+	srv.queue = make(chan *message, srv.QueueSize)
+	for i := 0; i < srv.Workers; i++ {
+		go func() {
+			for msg := range srv.queue {
+				srv.serve(msg)
+			}
+		}()
+	}
 }
 
 // ListenAndServe starts a nameserver on the configured address in *Server.
 func (srv *Server) ListenAndServe() error {
-	srv.lock.Lock()
-	defer srv.lock.Unlock()
-	if srv.started {
+	if srv.isRunning() {
 		return &Error{err: "server already started"}
 	}
+
+	if srv.Handler == nil {
+		srv.Handler = DefaultServeMux
+	}
```

Comment:
Why is this on needed?

*scrolls down*

Ah you're pulling it out from the serveX functions; sensible change; can you make that also a new PR?

Reply:
serveUDP() and serveTCP() have:
```
	handler := srv.Handler
	if handler == nil {
		handler = DefaultServeMux
	}
        ...
        go srv.serve(s.RemoteAddr(), handler, ...)
```
Why do we need to set handler each time on serveUDP() or serveTCP() call and then pass handler into serve() if we can set srv.Handler only once and serve() can just use srv.Handler?